src/app-layer-frames.c

```diff
@@ -0,0 +1,416 @@
+                return frame;
+        }
+    }
+    return NULL;
+}
+
+Frame *FrameGetByIndex(Frames *frames, const uint32_t idx)
+{
+    if (idx >= frames->cnt)
+        return NULL;
+
+    if (idx < FRAMES_STATIC_CNT) {
+        Frame *frame = &frames->sframes[idx];
+        FrameDebug("get_by_idx(s)", frames, frame);
+        return frame;
+    } else {
+        const uint16_t o = idx - FRAMES_STATIC_CNT;
+        Frame *frame = &frames->dframes[o];
+        FrameDebug("get_by_idx(d)", frames, frame);
+        return frame;
+    }
+}
+
+static Frame *FrameNew(Frames *frames, uint32_t rel_offset, int32_t len)
+{
+    BUG_ON(frames == NULL);
+
+    if (frames->cnt < FRAMES_STATIC_CNT) {
+        Frame *frame = &frames->sframes[frames->cnt];
+        frames->sframes[frames->cnt].rel_offset = rel_offset;
+        frames->sframes[frames->cnt].len = len;
+        frames->sframes[frames->cnt].id = frames->base_id++;
+        frames->cnt++;
+        return frame;
+    } else if (frames->dframes == NULL) {
+        BUG_ON(frames->dyn_size != 0);
+        BUG_ON(frames->cnt != FRAMES_STATIC_CNT);
+
+        frames->dframes = SCCalloc(8, sizeof(Frame));
+        if (frames->dframes == NULL) {
+            return NULL;
+        }
+        frames->cnt++;
+        BUG_ON(frames->cnt != FRAMES_STATIC_CNT + 1);
+
+        frames->dyn_size = 8;
+        frames->dframes[0].rel_offset = rel_offset;
+        frames->dframes[0].len = len;
+        frames->dframes[0].id = frames->base_id++;
+        return &frames->dframes[0];
+    } else {
+        BUG_ON(frames->cnt < FRAMES_STATIC_CNT);
+
+        /* need to handle dynamic storage of frames now */
+        const uint16_t dyn_cnt = frames->cnt - FRAMES_STATIC_CNT;
+        if (dyn_cnt < frames->dyn_size) {
+            BUG_ON(frames->dframes == NULL);
+
+            // fall through
+        } else {
+            if (frames->dyn_size == 256) {
+                SCLogNotice("limit reached! 256 dynamic frames already");
+                // limit reached
+                return NULL;
+            }
+
+            /* realloc time */
+            uint16_t new_dyn_size = frames->dyn_size * 2;
+            uint32_t new_alloc_size = new_dyn_size * sizeof(Frame);
+
+            void *ptr = SCRealloc(frames->dframes, new_alloc_size);
+            if (ptr == NULL) {
+                return NULL;
+            }
+
+            memset((uint8_t *)ptr + (frames->dyn_size * sizeof(Frame)), 0x00,
+                    (frames->dyn_size * sizeof(Frame)));
+            frames->dframes = ptr;
+            frames->dyn_size = new_dyn_size;
+        }
+
+        frames->cnt++;
+        frames->dframes[dyn_cnt].rel_offset = rel_offset;
+        frames->dframes[dyn_cnt].len = len;
+        frames->dframes[dyn_cnt].id = frames->base_id++;
+        return &frames->dframes[dyn_cnt];
+    }
+}
+
+static void FrameClean(Frame *frame)
+{
+    memset(frame, 0, sizeof(*frame));
+}
+
+static void StreamDPUCopy(Frame *dst, Frame *src)
```

Comment:
Q: Is this a typo or does it really mean something? `DPU`

Reply:
left over from older naming, will fix